lib/tree.rb

```diff
@@ -2,65 +2,150 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity:o(log n )
+  # Space Complexity: o(n)
   def add(key, value)
-    raise NotImplementedError
+    # key is what you are using to compare
+    # value is what you store
+    # if @root.nil?
+    #   @root = TreeNode.new(key, value)
+    # else
+    #   current = @root
+    #   while true
+    #     if key < current.key
+    #       if !current.left.nil?
+    #         current = current.left
+    #       else
+    #         current.left = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     else
+    #       if !current.right.nil?
+    #         current = current.right
+    #       else
+    #         current.right = TreeNode.new(key, value)
+    #         return
+    #       end
+    #     end
+    #   end
+    # end
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current_node, key, value)
+    # base case when the current node is nil
+    return TreeNode.new(key, value) if current_node.nil?
+    if key < current_node.key
+      current_node.left = add_helper(current_node.left, key, value)
+    else
+      current_node.right = add_helper(current_node.right, key, value)
+    end
+    return current_node
+  end
+
+  # Time Complexity:o(log n)
+  # Space Complexity:o(n)
   def find(key)
```

Comment:
The space/time complexity is O(log n) if the tree is balanced and O(n) if it's not.
The method however works